sdk/core/core-rest-pipeline/src/retryStrategies/failoverRetryStrategy.ts

```diff
@@ -0,0 +1,173 @@
+  return (retryableResponse || retryableError) as boolean;
+}
+
+/**
+ * Default behavior for generating hosts for failover.
+ *
+ * If the request method is GET, HEAD, or OPTIONS, it cycles through `readHosts`. Otherwise, it cycles through `writeHosts`.
+ */
+export function readWriteFailoverHostDelegate(options: {
+  readHosts?: string[];
+  writeHosts?: string[];
+}): FailoverHostDelegate {
+  const { readHosts, writeHosts } = options;
+  const readHostStateList: FailoverHosts = {
+    index: 0,
+    hostStates: (readHosts ?? []).map((url) => {
+      const host = new URL(url).host;
+      return { host, retryCount: 0 };
+    }),
+  };
+  const writeHostStateList: FailoverHosts = {
+    index: 0,
+    hostStates: (writeHosts ?? []).map((url) => {
+      const host = new URL(url).host;
+      return { host, retryCount: 0 };
+    }),
+  };
+
+  return function* (retryState, factoryOptions) {
+    // Optimization for the case where no retry hosts are specified
+    if (!readHostStateList.hostStates.length && !writeHostStateList.hostStates.length) {
+      return;
+    }
+
+    const retryDelayInMs = factoryOptions.retryDelayInMs ?? DEFAULT_CLIENT_RETRY_INTERVAL;
+    const maxRetryDelayInMs = factoryOptions.maxRetryDelayInMs ?? DEFAULT_CLIENT_MAX_RETRY_INTERVAL;
+
+    while (true) {
+      const request = retryState.response?.request || retryState.responseError?.request;
+      if (!request) {
+        throw Error("Request should be defined.");
+      }
+
+      // Select correct list of hosts for this request
+      const isSafeMethod = ["GET", "HEAD", "OPTIONS"].includes(request.method);
+      const currentStateList = isSafeMethod ? readHostStateList : writeHostStateList;
+
+      if (currentStateList.hostStates.length === 0) {
+        retryState = yield undefined;
+        continue;
+      }
+      const hostState = currentStateList.hostStates[currentStateList.index];
+
+      const requestFireTime = Math.max(hostState.retryAt?.valueOf() ?? -Infinity, Date.now());
+
+      retryState = yield hostState;
+
+      const exponentialDelay = _retryAfterUtil.exponentialDelayInMs(
+        hostState.retryCount,
+        retryDelayInMs,
+        maxRetryDelayInMs
+      );
+      const nextRetryAt = new Date(requestFireTime + exponentialDelay);
+      hostState.retryAt = nextRetryAt;
+
+      currentStateList.index = (currentStateList.index + 1) % currentStateList.hostStates.length;
+      hostState.retryCount++;
+    }
+  };
+}
+
+/**
+ *
+ * @param options - Lists of fallback endpoints for read/write operations
+ * @returns An exponential retry strategy with support for multiple endpoints
+ */
+export function failoverRetryStrategy(options: DefaultRetryPolicyOptions): RetryStrategy {
+  const delegateMap = new WeakMap<PipelineRequest, ReturnType<FailoverHostDelegate>>();
+  return {
+    name: "failoverRetryStrategy",
+    retry(state) {
+      if (!shouldFailover(state) || !options.failoverHostDelegate) {
```

Comment:
nit: putting the `!options.failoverHostDelegate` check first would be a little more performant due to short-circuiting

Reply:
Done!